sfge/src/main/java/com/salesforce/rules/multiplemassschemalookup/MultipleMassSchemaLookupVisitor.java

```diff
@@ -46,29 +44,47 @@ public void afterVisit(MethodCallExpressionVertex vertex, SymbolProvider symbols
             isSinkVisited = true;
             checkIfInsideLoop(vertex, symbols);
         } else if (RuleConstants.isSchemaExpensiveMethod(vertex) && shouldContinue()) {
-            violations.add(
-                    new MultipleMassSchemaLookupInfo(
-                            sourceVertex,
-                            sinkVertex,
-                            RuleConstants.RepetitionType.MULTIPLE,
-                            vertex));
+            createViolation(RuleConstants.RepetitionType.MULTIPLE, vertex);
         }
     }
 
     private void checkIfInsideLoop(MethodCallExpressionVertex vertex, SymbolProvider symbols) {
         final Optional<LoopBoundary> loopBoundaryOptional = loopBoundaryDetector.peek();
         if (loopBoundaryOptional.isPresent()) {
-            // Method has been invoked inside a loop. Create a violation.
-            final SFVertex loopVertex = loopBoundaryOptional.get().getBoundaryItem();
-            violations.add(
-                    new MultipleMassSchemaLookupInfo(
-                            sourceVertex,
-                            sinkVertex,
-                            RuleConstants.RepetitionType.LOOP,
-                            loopVertex));
+            if (!isLoopOutlier(vertex, symbols)) {
+                // Method has been invoked inside a loop. Create a violation.
+                final SFVertex loopVertex = loopBoundaryOptional.get().getBoundaryItem();
+                createViolation(RuleConstants.RepetitionType.LOOP, loopVertex);
+            }
         }
     }
 
+    private void createViolation(RuleConstants.RepetitionType type, SFVertex repetitionVertex) {
+        violations.add(
+                new MultipleMassSchemaLookupInfo(sourceVertex, sinkVertex, type, repetitionVertex));
+    }
+
+    /**
+     * Identifies cases where even if the call is within a loop, this action wouldn't be called
+     * multiple times.
+     *
+     * @param vertex Method call to examine
+     * @param symbols
+     * @return true if the method call would be called only once even if it's in a loop.
+     */
+    private boolean isLoopOutlier(MethodCallExpressionVertex vertex, SymbolProvider symbols) {
```

Comment:
The main change for this PR. I have another change coming up to exclude invocations from static code blocks.